### PR TITLE
Add support for JVM metrics reporting

### PR DESCRIPTION
**Pull request by shamsimam**
screen shot 2017-07-07 at 10 19 58 am

### CODE REVIEW

File: waiter/src/waiter/main.clj [hunk 1]
```diff
@@ -117,6 +118,7 @@
     (reify Thread$UncaughtExceptionHandler
       (uncaughtException [_ thread throwable]
         (log/error throwable (str (.getName thread) " threw exception: " (.getMessage throwable))))))
+  (jvm-metrics/instrument-jvm)
```
Comment (dposada) [inline waiter/src/waiter/main.clj]: Does this have an effect on performance?

Comment (pschorf) [inline waiter/src/waiter/main.clj]: 👍

Comment (shamsimam) [inline waiter/src/waiter/main.clj]: I haven't measured the performance impact but wouldn't expect these to be high. The jvm metrics are all gauges, which means they only execute when we ask for the value.

**Comment by shamsimam** [comment]
any more feedback on this?

**Comment by dposada** [comment]
Should we add an integration test for this?

**Comment by shamsimam** [comment]
@dposada Yeah, I'll add checking the presence of "jvm" entries.

File: waiter/src/waiter/main.clj (hunk 1 above)
Comment (shamsimam) [inline waiter/src/waiter/main.clj]: Inter-router metrics sync-ing is triggered via `waiter.metrics/get-core-metrics` which uses a `MetricFilter` that looks into service metrics only.

File: waiter/integration/waiter/metrics_output_test.clj [hunk 2]
```diff
@@ -58,9 +64,37 @@
                               s/Str s/Any}
    s/Str s/Any})
 
+(def jvm-metrics-schema
+  {
+   (s/required-key "attribute") s/Any
```
Comment (dposada) [inline waiter/integration/waiter/metrics_output_test.clj]: Shouldn't this be on the previous line?

File: waiter/integration/waiter/metrics_output_test.clj [hunk 3]
```diff
@@ -58,9 +64,37 @@
...
+                              (s/required-key "runnable") counted-gauge-metric-schema
+                              (s/required-key "terminated") counted-gauge-metric-schema
+                              (s/required-key "timed_waiting") counted-gauge-metric-schema
+                              (s/required-key "waiting") counted-gauge-metric-schema
+                              s/Str s/Any}
+   s/Str s/Any})
+
+(def waiter-metrics-schema
+  {
+   (s/required-key "autoscaler") s/Any
```
Comment (dposada) [inline waiter/integration/waiter/metrics_output_test.clj]: Shouldn't this be on the previous line?

File: waiter/integration/waiter/metrics_output_test.clj [hunk 4]
```diff
@@ -83,11 +117,15 @@
...
-                      (log/info "Asserting /metrics output for" router-url)
-                      (assert-metrics-output app-metrics)))
+                          metrics-json-response (make-request router-url "/metrics")
+                          metrics-response (json/read-str (:body metrics-json-response))
+                          service-metrics (get-in metrics-response ["services" service-id])]
+                      (log/info "Asserting jvm metrics output for" router-url)
+                      (assert-metrics-output (get metrics-response "jvm") jvm-metrics-schema)
+                      (log/info "Asserting service metrics output for" router-url)
+                      (assert-metrics-output service-metrics service-metrics-schema)
+                      (log/info "Asserting waiter metrics output for" router-url)
```
Comment (dposada) [inline waiter/integration/waiter/metrics_output_test.clj]: Nitpick: lowercase first word of log statements

File: waiter/src/waiter/main.clj (hunk 1 above)
Comment (dposada) [inline waiter/src/waiter/main.clj]: So just calling `(jvm-metrics/instrument-jvm)` doesn't add any overhead?

File: waiter/src/waiter/handler.clj [hunk 5]
```diff
@@ -191,11 +191,13 @@
   [request]
   (try
     (let [request-params (:params (ring-params/params-request request))
-          exclude-services (Boolean/parseBoolean (get request-params "exclude-services" "false"))
+          exclude-services (utils/request-flag request-params "exclude-services")
           service-id (get request-params "service-id" nil)
+          show-jvm-metrics (utils/request-flag request-params "show-jvm-metrics")
```
Comment (sradack) [inline waiter/src/waiter/handler.clj]: Should this be named `include-jvm-metrics` to be consistent with `exclude-services`?

File: waiter/src/waiter/main.clj (hunk 1 above)
Comment (sradack) [inline waiter/src/waiter/main.clj]: `instrument-jvm` creates gauges.  The overhead comes when the gauges are queried, at which point each gauge calls its backing function to return the current value.

Comment (sradack) [inline waiter/src/waiter/main.clj]: And based my understanding, the only time we ask for the value of the jvm gauges is in the handler.  We do not evaluate them automatically on a regular basis.

**Comment by shamsimam** [comment]
@dposada 

> So just calling (jvm-metrics/instrument-jvm) doesn't add any overhead?

`instrument-jvm` ends up creating gauges, the overhead with gauges only occurs when we try to extract their value (which we aren't during this call).

**Comment by sradack** [comment]
Aside from @dposada's comments, LGTM

**Comment by shamsimam** [comment]
@dposada responded to your comments. Passing build: https://travis-ci.org/twosigma/waiter/builds/252550413

**Comment by shamsimam** [comment]
@pschorf any more feedback on this?

**Comment by shamsimam** [comment]
Can we get this merged?

**Pull request merged**